internal/sdkprovider/kafkatopicrepository/repository_test.go

```diff
@@ -0,0 +1,285 @@
+				{topic: &aiven.KafkaTopic{TopicName: "c"}},
+				{topic: &aiven.KafkaTopic{TopicName: "a"}},
+				{topic: &aiven.KafkaTopic{TopicName: "b"}},
+				{topic: &aiven.KafkaTopic{TopicName: "c"}},
+				{topic: &aiven.KafkaTopic{TopicName: "a"}},
+				{topic: &aiven.KafkaTopic{TopicName: "b"}},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/a/a": {TopicName: "a"},
+				"a/a/b": {TopicName: "b"},
+				"a/a/c": {TopicName: "c"},
+				"a/b/a": {TopicName: "a"},
+				"a/b/b": {TopicName: "b"},
+				"a/b/c": {TopicName: "c"},
+				"b/a/a": {TopicName: "a"},
+				"b/a/b": {TopicName: "b"},
+			},
+			v1ListCalled: 3, // 3 different cervices
+			// 2 services has 3 topics each.
+			// Plus one service has 2 topics
+			// Gives us batches (brackets) with topics (in brackets):
+			// [2] + [1] + [2] + [1] + [2]
+			v2ListCalled:    5,
+			v2ListBatchSize: 2,
+		},
+		{
+			name: "v1List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("bla bla bla")},
+			},
+			v1ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    0,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: bla bla bla")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List 404 error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: topic list has changed")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       aiven.Error{Status: 404},
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+	}
+
+	for _, opt := range cases {
+		t.Run(opt.name, func(t *testing.T) {
+			client := &fakeTopicClient{
+				storage:   opt.storage,
+				v1ListErr: opt.v1ListErr,
+				v2ListErr: opt.v2ListErr,
+			}
+
+			ctx := context.Background()
+			rep := newRepository(client)
+			rep.workerCallInterval = time.Millisecond
+			rep.v2ListBatchSize = opt.v2ListBatchSize
+
+			// We must run all calls in parallel
+			// and then call the worker
+			var wg sync.WaitGroup
+			for i := range opt.requests {
+				wg.Add(1)
+				go func(i int) {
+					defer wg.Done()
+					topic, err := rep.Read(ctx, opt.requests[i].project, opt.requests[i].service, opt.requests[i].topic)
+					assert.Equal(t, opt.responses[i].topic, topic)
+					assert.Equal(t, opt.responses[i].err, err)
+				}(i)
+			}
+
+			go rep.worker()
```

Comment:
consider ensuring there is a mechanism to cleanly shutdown these goroutines

Reply:
It is a test case. Not a running service app. Test is done, goroutine is removed.